packages/liveblocks-node/src/thread-notification-helpers.ts

```diff
@@ -0,0 +1,284 @@
+ * {
+ *  type: "unreadReplies",
+ *  comments: [unread_comment_0, unread_comment_1, ...],
+ * }
+ *
+ * @example Unread mention:
+ * {
+ *  type: "unreadMention",
+ *  comments: [unread_comment_with_mention]
+ * }
+ */
+export async function getThreadNotificationData(params: {
+  client: Liveblocks;
+  event: ThreadNotificationEvent;
+}): Promise<ThreadNotificationData> {
+  const { client, event } = params;
+  const { threadId, roomId, userId, inboxNotificationId } = event.data;
+
+  const [thread, inboxNotification] = await Promise.all([
+    client.getThread({ roomId, threadId }),
+    client.getInboxNotification({ inboxNotificationId, userId }),
+  ]);
+
+  const readAt = inboxNotification.readAt;
+  const unreadComments = thread.comments
+    .filter((c) => c.userId !== userId)
+    .filter((c) => c.body !== undefined && c.deletedAt === undefined)
+    .filter((c) =>
+      readAt
+        ? c.createdAt > readAt && c.createdAt <= inboxNotification.notifiedAt
+        : c.createdAt <= inboxNotification.notifiedAt
+    ) as ThreadNotificationCommentData[];
+
+  const lastUnreadCommentWithMention = getLastCommentWithMention({
+    comments: unreadComments,
+    mentionedUserId: userId,
+  });
+
+  if (lastUnreadCommentWithMention !== null) {
+    return {
+      type: "unreadMention",
+      comments: [lastUnreadCommentWithMention],
+    };
+  }
+
+  return {
+    type: "unreadReplies",
+    comments: unreadComments,
+  };
+}
+
+export type ResolveRoomInfoArgs = {
+  /**
+   * The ID of the room to resolve
+   */
+  roomId: string;
+};
+
+export type GetThreadNotificationUnreadCommentsDataOptions<
+  U extends BaseUserMeta = DU,
+> = {
+  /**
+   * Which format to transform the comment body to.
+   */
+  format?: "html" | "json";
+  /**
+   * A function that returns info from user IDs.
+   */
+  resolveUsers?: (
+    args: ResolveUsersArgs
+  ) => OptionalPromise<(U["info"] | undefined)[] | undefined>;
+  /**
+   * A function that returns room info from room IDs.
+   */
+  resolveRoomInfo?: (
+    args: ResolveRoomInfoArgs
+  ) => OptionalPromise<DRI | undefined>;
+};
+
+export type UnreadCommentAuthorData = {
+  id: string;
+  name: string;
+  avatar?: string;
+};
+
+/** @internal */
+export const getAuthor = async <U extends BaseUserMeta = DU>({
+  userId,
+  resolveUsers,
+}: {
+  userId: string;
+  resolveUsers?: (
+    args: ResolveUsersArgs
+  ) => OptionalPromise<(U["info"] | undefined)[] | undefined>;
+}): Promise<UnreadCommentAuthorData> => {
+  const fallback: UnreadCommentAuthorData = {
+    id: userId,
+    name: userId,
+  };
+  if (resolveUsers) {
+    const users = await resolveUsers({ userIds: [userId] });
+    return users?.[0] && users?.[0].name
+      ? { id: userId, name: users[0].name }
+      : fallback;
+  }
+  return fallback;
+};
+
+export type UnreadCommentData = {
+  id: string;
+  threadId: string;
+  roomId: string;
+  author: UnreadCommentAuthorData;
+  createdAt: Date;
+  body: string | CommentBodyJson;
```

Comment:
This might be hard to work with? Ideally the function would return the type you asked for

Reply:
Yes nice catch here 👍🏻 The type should return either a body of type `string` or either of type `json` 👍🏻